src/audio/copier/copier.c

```diff
@@ -1720,18 +1720,27 @@ static uint64_t copier_get_processed_data(struct comp_dev *dev, uint32_t stream_
 {
 	struct copier_data *cd = comp_get_drvdata(dev);
 	uint64_t ret = 0;
-	bool source = dev->direction == SOF_IPC_STREAM_CAPTURE;
 
-	if (cd->endpoint_num && cd->bsource_buffer != input) {
-		if (stream_no < cd->endpoint_num) {
-			if (dev->ipc_config.type == SOF_COMP_HOST && !cd->ipc_gtw) {
-				if (source == input)
-					ret = cd->hd->total_data_processed;
-			} else {
-				ret = comp_get_total_data_processed(cd->endpoint[stream_no],
-								    0, input);
-			}
+	/*
+	 * total data processed is calculated as: accumulate all input data in bytes
+	 * from host, then store in host_data structure, this function intend to retrieve
+	 * correct processed data number.
+	 * Due to host already integrated as part of copier, so for host specific case,
+	 * as long as direction is correct, return correct processed data.
+	 * Dai still use ops driver to get data, later, dai will also be integrated
+	 * into copier, this part will be changed again for dai specific.
+	 */
+	if (cd->endpoint_num) {
+		if (dev->ipc_config.type == SOF_COMP_HOST && !cd->ipc_gtw) {
+			bool source = dev->direction == SOF_IPC_STREAM_PLAYBACK;
+
+			if (source == input)
+				ret = cd->hd->total_data_processed;
+		} else {
+			ret = comp_get_total_data_processed(cd->endpoint[stream_no],
```

Comment:
we're now not evaluating `stream_no` at all in this function and here we're using it as an index into an array. I don't know what the original intention of this parameter was. I see that it's currently always 0. But maybe we should put something like
```
if (stream_no >= cd->endpoint_num) {
	comp_err();
	return 0;
}
```
between lines 1733 and 1734 above?

Reply:
same result with:
if (stream_no < cd->endpoint_num) 

Since we are not able to see stream_no >= cd->endpoint_num, I removed this condition.

if need add, then we can add above code back, right? add above code back also have return 0, since ret initialized as 0.

Reply:
yes, this achieves the same except it isn't printing an error